recitation_3/memorypointers.md

```diff
@@ -0,0 +1,282 @@
+
+```c
+int foo()
+{
+    static int x = 5; // assign value of 5 only once
+    x++;
+    return x;
+}
+
+int main()
+{
+    printf("%d\n", foo()); // prints 6
+    printf("%d\n", foo()); // prints 7
+    return 0;
+}
+```
+
+A **global variable** is like a special case of static variables. It is accessible from all files in the program, and can be accessed from other files using the `extern` keyword. See below:
+
+In one file:
+
+```c
+int global_static = 0;
+
+int main() {
+    global_static++;
+    magic_print();
+	return 0;
+}
+```
+
+In another file:
+
+```c
+void magic_print() {
+    extern int global_static;
+    printf("%d\n", global_static); // prints 1
+}
+```
+
+## Pointers ##
+
+Understanding pointers is crucial and using them naturally will make you stand out as a programmer. Let's start with
+the basics.
+
+**Pointer:** A variable that stores a memory address. That's it. The memory address is _just a number_.
+
+There isn't just one data type called "pointer". Every pointer is a
+pointer-to-type, which encodes how to interpret the bytes you find in the memory
+address.
+
+In a variable declaration, an asterisk denotes the fact that the variable is a
+pointer.
+
+Here's an example:
+
+```c
+int x = 5; // x is a plain int
+int *p; // p is a pointer-to-int
+```
+
+### Using basic pointers ###
+
+There are two basic operators that you use with pointers: 
+
+The `&` operator references a value, ie it gets the memory address of an
+already existing variable. It could then be stored into a pointer.
+
+```c
+int x = 5; // x is a plain int
+int *p; // p is a pointer-to-int
+p = &x; // p now points to x
+```
+
+The `*` operator *dereferences* a pointer: it follows the pointer to retrieve the value it points to
+
+```c
+int x = 5;
+int *p = &x; 
+printf("%d", *p); //prints out 5
+*p = 9;  // now x is 9
+```
+
+`&` and `*` are basically opposites: `&` adds a level of indirection, while `*` removes a level, bringing you
+closer to the underlying value. 
+
+
+### Ok, so why use pointers? ###
+
+C is a **call-by-value** language which means all arguments to functions are 
+copied, and a local copy is made on that function's stack. Changes made inside the
+function are not reflected on the outside. 
+__Therefore if you want a function to
+modify a value that you have, you'll have to tell the function where to find the
+that value by memory address, not just give it the value:__
+
+```c
+void increment(int a) {
+  a++;
+}
+
+void actually_increment(int *b) {
+  (*b)++;
+}
+
+int main() {
+  int x = 1;
+  increment(x); // x is still 1
+  actually_increment(&x); // x is now 2
+  return 0;
+}
+```
+
+Note not only the difference in the function, but how the parameters are passed.
+**Passing a pointer is fundamentally a different type than passing a value.**
```

Comment:
Phrasing is weird here

Reply:
i agree, it is a bit vague what it means.

Reply:
It may be helpful to elaborate on explanation. For ex: int a is a copy of int x, that it only "lives" in increment() and is trashed afterwards. On the other hand *b goes to the actual int x and modifies that.